### ISSUE TITLE
Allow prefixed names to be used as the endpoint in a SERVICE clause

### ISSUE THREAD

**Issue by mgberg** [RDFLib/rdflib#1427]
Federated queries do not work when the provided endpoint in a SERVICE clause is a prefixed name. I have tracked the error down to the [`evalServiceQuery` function](https://github.com/RDFLib/rdflib/blob/038af4547e799845e08986867b24acafcbe72d48/rdflib/plugins/sparql/evaluate.py#L302).

The regular expression used (`"^service <(.*)>[ \n]*{(.*)}[ \n]*$"`) will not match a prefixed name so the query will not be evaluated. That regex should be adjusted to include prefixed names.

Also, note that if a prefixed name is currently used as the endpoint, `part.term` already contains the correctly expanded `URIRef` of the endpoint that should be able to be used as the `service_url` a few lines later; the prefixed name matched by the regular expression shouldn't need to be expanded.

**Comment by shashankdaima** [comment]
can you assign this to me? @mgberg

**Comment by shashankdaima** [comment]
@mgberg can you please give me the example of query that will cause error?

**Comment by mgberg** [comment]
In rdflib 6.1.1, the following works as expected:
```python
list(
    g.query("""\
        SELECT *
        WHERE {
          SERVICE <https://dbpedia.org/sparql> {
            {
              SELECT ?s ?p ?o
              WHERE {
                ?s ?p ?o
              }
              LIMIT 10
            }
          }
        }"""
    )
)
```
However, the following does not work:
```python
g.bind("dbp", "https://dbpedia.org/")
list(
    g.query("""\
        SELECT *
        WHERE {
          SERVICE dbp:sparql {
            {
              SELECT ?s ?p ?o
              WHERE {
                ?s ?p ?o
              }
              LIMIT 10
            }
          }
        }"""
    )
)
```

**Comment by shashankdaima** [comment]
@mgberg  and prefix can entered like this as well? ```PREFIX dbp: <http://dbpedia.org/>
        SELECT *
        WHERE {
          SERVICE dbp:sparql {
            {
              SELECT ?s ?p ?o
              WHERE {
                ?s ?p ?o
              }
              LIMIT 1
            }
          }
        }```.

**Comment by mgberg** [comment]
Explicitly specifying the prefix does not help. This snippet doesn't work for me either: g.bind("dbp", "https://dbpedia.org/")
list(
    g.query("""\
        PREFIX dbp: <http://dbpedia.org/>
        SELECT *
        WHERE {
          SERVICE dbp:sparql {
            {
              SELECT ?s ?p ?o
              WHERE {
                ?s ?p ?o
              }
              LIMIT 10
            }
          }
        }"""
    )
)

**Comment by shashankdaima** [comment]
yes thanks. can you tell me how can i get prefix map? is there any hashmap/dict for prefixes?

**Comment by mgberg** [comment]
I don't think a prefix map is necessary. As I pointed out, in `evalServiceQuery`, `part.term` already contains the correctly expanded `URIRef` of the endpoint that should be able to be used as the `service_url`. You can see this by running the following snippet:
```python
from rdflib.plugins.sparql import prepareQuery
from rdflib import URIRef

query_text = """\
PREFIX dbp: <http://dbpedia.org/>
SELECT *
WHERE {
  SERVICE dbp:sparql {
    {
      SELECT ?s ?p ?o
      WHERE {
        ?s ?p ?o
      }
      LIMIT 10
    }
  }
}"""

prepared_query = prepareQuery(query_text)
part = prepared_query.algebra.p.p   # This essentially becomes `part` in `evalServiceQuery`
assert part.term == URIRef("http://dbpedia.org/sparql")
```
Here you can see that the prefix expansion has already been performed. If you put a breakpoint in `evalServiceQuery` and inspect `part`, you'll see the same thing. It's just a matter of changing how `evalServiceQuery` tries to get a service URL, which might include looking at `part.term` instead of using a regular expression.

As I did not write the query processor and was not involved with its development in any way, I was apprehensive about making a pull request myself as I didn't want to break anything.

**Comment by gjhiggins** [comment]
> I was apprehensive about making a pull request myself to fix this issue because I didn't want to make a change based on bad assumptions.

Ohh, I wouldn't worry about *that*, I do it all the time (not intentionally :smile:) The maintenance team's primary constraint is time and giving a bit of guidance here and there is generally less demanding that having to do it all oneself. And, once you've surmounted the contribution barrier, it is definitely easier the next time. *All* help with improving/fixing/maintaining RDFLib is very much appreciated.

And anyway, now you've expanded on your diagnosis, fwiw I'd say you were correct both in identifying a likely fix and being cautious about context of federated queries. Happily, this can all be tested in a reasonably straightforward manner.
